### ISSUE TITLE
DI and Community Toolkit > 

### ISSUE THREAD

**Issue by kscavitt** [VsixCommunity/Community.VisualStudio.Toolkit.DependencyInjection#7]
Hello,

Not sure if this should be in this repo or the main Community.VisualStudio.Toolkit repo issue list.

I'm seeing dependency issues when trying to update a vsix package I created to use the latest version of Community.VisualStudio.Toolkit (17.0.475) 

From my activity.xml file.
![image](https://user-images.githubusercontent.com/68757574/196491239-a551372c-adb3-4467-9bce-ed26ca8983d5.png)

I was looking at the Dependencies and notice that the Community,VisualStudio.DependencyInjection.Core has a dependency on Community.VisualStudio.Toolkit >= 17.0.430. I assumed that nuget was choosing the lowest dependency and converted the project to use a package.config and the DependencyVersion Highest in my Nuget.Config, but my package still fails to load..

I can see the latest version in my vsix file if I rename and unzip the file.
![image](https://user-images.githubusercontent.com/68757574/196494476-cec19ed1-8326-4ef0-a8e1-511afe63a512.png)

Test Repo:
https://github.com/kscavitt/VsixExtensionTest/

**Comment by StevenRasmussen** [comment]
Hello, what version of Visual Studio are you using to test this?  I just ran the code in your test repo in VS 2022 and it appeared to work without any issues.  If you are developing an extension for VS 2019 (and developing in VS 2019) then you will need to update the references to use the v16.x NuGet packages instead of v17.  The versions of the NuGet packages correlate with the minimum version of Visual Studio that you are targeting.

**Comment by kscavitt** [comment]
> Hello, what version of Visual Studio are you using to test this? I just ran the code in your test repo in VS 2022 and it appeared to work without any issues. If you are developing an extension for VS 2019 (and developing in VS 2019) then you will need to update the references to use the v16.x NuGet packages instead of v17. The versions of the NuGet packages correlate with the minimum version of Visual Studio that you are targeting.

VS 2022 version 17.3.6

You were able to run the extension in my linked project? It builds just fine. It's when I try to debug it that I get and error that points me to the ActivityLog.xml file in the experimental versions of VS 2022

**Comment by kscavitt** [comment]
> > You were able to run the extension in my linked project? It builds just fine. It's when I try to debug/run it that I get an error that points me to the ActivityLog.xml file in the experimental versions of VS 2022.
> 
> I was able to debug it and did not receive any errors.

This is totally strange. I just setup a new machine with VS 2022 and followed the installation [setup guide](https://learn.microsoft.com/en-us/visualstudio/extensibility/vsix/get-started/get-tools?view=vs-2022) and I get the same results when I try to launch my extension from View>Other Windows>My Tool Window

![image](https://user-images.githubusercontent.com/68757574/196592180-e8e3b1bb-df23-43b4-a39d-484f07f24afe.png)

Checking the xml file. It failed to find Community.VisualStudio.Toolkit 17.0.430.

**Comment by StevenRasmussen** [comment]
I never tried opening up the tool window. Is that when you’re getting the error? I can try that tomorrow and see what happens for me.

**Comment by kscavitt** [comment]
> I never tried opening up the tool window. Is that when you’re getting the error? I can try that tomorrow and see what happens for me.

Yes. I get the error when I try to load the tool window. This project is just the basic template that come with the [Extensibility Essentials 2022](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.ExtensibilityEssentials2022) install but modified to use dependency injection via Community.VisualStudio.Toolkit.DependencyInjection. 

![image](https://user-images.githubusercontent.com/68757574/196597675-6ab1becd-fe02-40f3-b06c-de38fe985feb.png)


It should show the Click Me button. (it works fine with Community.VisualStudio.Toolkit 17.0.430.0) 
![image](https://user-images.githubusercontent.com/68757574/196598303-dfd1d934-96fe-4987-b014-c2b96c309c93.png)

**Comment by KC250330** [comment]
I thought I should probably add the steps to reproduce.

# Steps to Reproduce

1. Create a new project using the VSIX Project w/Tool Window (Community) template called VSIXProject1.
2. Add the following Nuget packages.
    a. Community.VisualStudio.Toolkit.DependencyInjection.Core.17
    b. Community.VisualStudio.Toolkit.DependencyInjection.Microsoft.17
3. Modify the MyToolWindowCommand.cs file to inherit from **BaseDICommand** instead of BaseCommand<MyToolWindowCommand>.
4. Modify the VSIXProject1Package.cs.
    a. Inherit from **MicrosoftDIToolkitPackage<VSIXProject1Package>** instead of ToolkitPackage
    b. Remove the current **InitializeAsync** method and replace it with the following.

              protected override void InitializeServices(IServiceCollection services)
              {
                  // Register commands.
                  services.AddSingleton<MyToolWindowCommand>();

                  // Register all commands in the assembly.
                  services.RegisterCommands(ServiceLifetime.Singleton);
               }

              protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
              {
                  // Ensure that you first call the base.InitializeAsync method.
                  await base.InitializeAsync(cancellationToken, progress);

              }
5. Update the Nuget package Community.VisualStudio.Toolkit to anything above 17.0.430, In my case I updated to the latest.(17.0.475)
6. Run in an experimental VS 2022 instance.
7. Open the tool window View>Other Windows>My Tool Windows

You will see a message similar to the following.
![image](https://user-images.githubusercontent.com/77290693/197249147-8035906d-2474-4031-87b7-199e93cadebc.png)

If you browse to the path of the ActivityLog.xml file listed in the similar message above, open it in edge, scroll to the bottom and you should see something like the picture below.

![image](https://user-images.githubusercontent.com/77290693/197249489-650cd54d-6c31-4a8e-8c28-21f85990e8c6.png)

**Comment by StevenRasmussen** [comment]
@KC250330 - I was able to reproduce the issue and it appears that it comes down to an issue with binding redirection.  The way that Visual Studio loads the extensions prevents it from using automatic binding redirection. See [here](https://stackoverflow.com/questions/63353365/binding-redirect-not-effective-in-vspackage-vsix#:~:text=It%20seems%20that%20the%20answer%20is%20that.NET%20was,own%20devenv.exe.config%20file%20that%20the%20VSIX%20can%27t%20edit.).  I was able to overcome this by adding the attribute as described in that SO question to the assembly.cs file:
```csharp
[assembly: ProvideBindingRedirection(AssemblyName = "Community.VisualStudio.Toolkit",
    NewVersion = "17.0.475.0", OldVersionLowerBound = "17.0.430.0", OldVersionUpperBound = "17.0.451.0")]
```

One thing I did notice is that it appears you have a lot of unnecessary references to VS SDK related packages in that project. You should only need to reference the `Community.VisualStudio.Toolkit` package and the DI package.

**Comment by KC250330** [comment]
@StevenRasmussen  - Thanks for the investigation. That fixed the "Community.VisualStudio.Toolkit" version issue. However it's now complaining about "Microsoft.Extensions.DependencyInjection".  

Do you happen to know of a guide that describes how to figure out what the binding redirect versions should look like? More specifically how to find the NewVersion, OldVersionLowerBound, OldVersionUpperBound?  I would like to figure out how to address these types of issues in the future.

![image](https://user-images.githubusercontent.com/77290693/197547549-7435f6c2-7466-404c-8fe3-a8bae6ecaeb2.png)

**Comment by StevenRasmussen** [comment]
I unfortunately don't have anything as far as a guide goes for determining what the binding redirect should look like.  What I did for the one above is simply look at what version of the package was installed and used that as the `NewVersion`.  I then just picked the version for the package that was expected as the `OldVersionLowerBound` (this is in the error message). Then, looking at what versions are available for the package, I picked the version right below what was installed as the `OldVersionUpperBound`.

Regarding this error in general:  There appears to be two ways to solve it:
1. Install the version that the error message tells you is not able to be found (this is probably the preferred solution and may work for the `Microsft.Extensions.DependencyInjection` error you're now facing).
2. Or, add a binding redirect attribute for the version that is installed.

Hopefully that helps :)
